### PR TITLE
Add tool for generating a normalized text representation of mark/kerning rules

### PR DESCRIPTION
**Pull request by cmyr**
This adds a new tool for generating a normalized representation of various rules rules in the GPOS table (currently pairpos, mark-base, and mark-mark), and wires this up to the `ttx_diff` utility.

When you run `ttx_diff`, we will now generate two additional files, `fontc.markkern.txt` and `fontmake.markkern.txt`, in addition to the ttx. These files can be inspected to provide more insight into what exactly we're compiling.

Important distinctions:
- we do not include any info about lookups. For each feature/lang/script combination, we print all the rules referenced in any of the handled lookups for that feature, sorted by lookup type and then generally in gid order. This discards important information, but it is consistent
- ditto subtables
- we resolve deltas, and print them inline.
- we print lookupflags and filter glyph sets inline, anytime a rule has flags/filter glyphs that differ from the last printed rule
- we will print lots of duplicates! if a lookup is referenced by multiple features, we print its rules for each one.

stuff that's bad
- the name of the crate, probably?
- the files we produce are very big
- needs more docs
- there's some fancy use of `dyn trait` that might make some people mad at me at some point

anyway: I have mixed feelings about this work, but ultimately it is *useful*; in the process of writing this and debugging its output I have developed a much better understanding of what's going on in our compilation of these tables, and I was able to identify a number of bugs, so I think this is worth having.

this currently includes the commit from #577, but that should be removed before we commit.

### CODE REVIEW

File: layout-repr/build.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+//! Build script to generate our glyph name lookup table.
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply layout-repr/build.rs:1]: Given our use case what if we give you a glyph order and we delete the aglfn.txt bit entirely?

**Review by rsheeter — APPROVED** [review]
Great start, lets merge and iterate

**Comment by anthrotype** [comment]
the variable anchors notation is a bit hard to parse. Instead of

> @(246 {234, 275},0 {0, 0})

how about grouping the (x, y) tuples together

> A @((246, 0) {(234, 0), (275, 0)})

File: resources/scripts/ttx_diff.py [hunk 2]
```diff
@@ -46,17 +46,21 @@
...
 
 
-def run(cmd: MutableSequence, working_dir: Path, log_file: str, **kwargs):
+def run(cmd: MutableSequence, working_dir: Path, log_file: str, redirect_stdout: Optional[str] = None, **kwargs):
     cmd_string = " ".join(cmd)
     print(f"  (cd {working_dir} && {cmd_string} > {log_file} 2>&1)")
     log_file = working_dir / log_file
     with open(log_file, "w") as log_file:
+        if redirect_stdout is not None:
+            stdout = open( working_dir / redirect_stdout, "w")
```
Comment (anthrotype) [inline resources/scripts/ttx_diff.py:55]: pls re-run `black` autoformatter, the stray space after `open( `.

File: resources/scripts/ttx_diff.py [hunk 3]
```diff
@@ -70,10 +74,20 @@ def ttx(font_file: Path):
...
     ]
-    run(cmd, font_file.parent, "ttx.log")
+    run(cmd, font_file.parent, "ttx.log", None)
     return ttx_file
 
+# generate a simple text repr for gpos for this font
+def simple_gpos_output(font_file: Path, out_path: Path):
+    temppath = font_file.parent / "gpos.txt"
+    cmd = ["cargo", "run", "-p", "layout-repr", "--", font_file.name, "--table", "gpos"]
+    run(cmd, font_file.parent, "gpos.log", temppath.name,)
```
Comment (anthrotype) [inline resources/scripts/ttx_diff.py:84]: maybe instead of `temppath.name` you could pass an open file object to redirect stdout to, and also make sure that you close the file after you finish writing to it, before you do the copy (this is good practice in general, and windows is very sensitive file handles left open). 

```python
with open(temppath.name, "w") as tempfile:
    run(cmd, font_file.parent, "gpos.log", tempfile)
```

File: layout-repr/build.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmyr) [reply layout-repr/build.rs:1]: > Other parts of fontc that generate tables like this use python, e.g. `resources/scripts/non_spacing_marks.py`. It would be nice if this was consistent.

That's fair, but I'd prefer not to take this working build script (which is, in any case, the canonical way to do this sort of thing in a rust project) and port it to python right now. I'm happy to open an issue to look into standardizing this sort of thing, though.


> Given our use case what if we give you a glyph order and we delete the aglfn.txt bit entirely?

I presume you mean a glyph order that we take from the source? My thinking for this tool was that it should work entirely on a font binary, without assuming access to a source, but this isn't 100% necessary.

I _suspect_ that on a long enough time frame we are going to want a way to get sensible names for glyphs in a binary font, and I _suspect_ that this will end up being a standalone crate that lives in `fontations` or elsewhere, but I don't have strong feelings here.

Given that this code exists and works I'd lean towards just leaving it for now? Are there any advantages to removing it that I might be overlooking?

File: layout-repr/src/common.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    pub(crate) fn combine(&mut self, other: GlyphSet) {
+        self.make_set();
+        let GlyphSet::Multiple(gids) = self else {
+            unreachable!()
+        };
+        match other {
+            GlyphSet::Single(gid) => gids.push(gid),
+            GlyphSet::Multiple(multi) => gids.extend(multi),
```
[the remark this answers is not in the document]
Comment (cmyr) [reply layout-repr/src/common.rs:113]: Good point, I think initially I wanted to preserve ordering or something but this is easier if I just use a BTreeSet internally.

File: resources/scripts/ttx_diff.py (hunk 3 above)
Comment (cmyr) [reply resources/scripts/ttx_diff.py]: I've reworked this all so that the normalizer tool can just take a `-o` option to write to a file, and we don't need to worry about this weird stdout redirection hack.

**Comment by cmyr** [comment]
Okay, I've done a bunch of cleanup and hopefully addressed most concerns, this should be ready to go in and we can iterate from here.

**Comment by cmyr** [comment]
Okay I'm going to merge this as-is (into master, without mark-to-mark working) and then I will fix mark-to-mark compilation afterwards.
